### PR TITLE
[CLIPPER-137] [CLIPPER-155] [CLIPPER-158] Deploy Scala Spark models no repl support

### PR DESCRIPTION
**Pull request by dcrankshaw**
Waiting on a few things:

- [x] Tests
- [ ] Documentation
- [x] #121 will create massive conflicts with this PR, so let's wait to even review until that PR is merged

Creating the pull request so I get a nice diff and can start to review it myself.

### CODE REVIEW

File: containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/container/NoOpStringModel.java
Comment (Corey-Zumar) [inline containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/container/NoOpStringModel.java]: Clean up imports. The following are unused:

```
import javax.xml.soap.SAAJResult;
import java.nio.FloatBuffer;
import java.util.Iterator;
import ai.clipper.container.data.FloatVector;
```

File: containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala]: Remove commented import

File: containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala]: Clean commented code

**Comment by dcrankshaw** [comment]
To test this:

1. Build example spark job: `cd <clipper-root>/containers/jvm && mvn clean package`
2. Download [Spark 2.1](http://spark.apache.org/downloads.html)
3. Build the Spark Container docker image: `cd <clipper-root>/containers/jvm && docker build -t clipper/spark-scala-container -f SparkScalaContainerDockerfile ./`
4. Start Clipper
5. Run the example Spark job using Spark submit : `CLIPPER_MODEL_NAME=<model-name> CLIPPER_HOST=<host> <path-to-spark>/bin/spark-submit --master "local[2]" --class ai.clipper.examples.train.Train --name <spark-job-name> <path-to-clipper>/containers/jvm/clipper-examples/target/clipper-examples-0.1.jar`

File: containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-examples/src/main/java/ai/clipper/examples/train/SparkTrainDeployExample.scala:92]: Let's print the test error as a percentage like we do with the Python examples - seems more meaningful than a quantity of incorrect predictions

File: containers/jvm/clipper-rpc/src/main/java/ai/clipper/rpc/RPC.java
Comment (Corey-Zumar) [inline containers/jvm/clipper-rpc/src/main/java/ai/clipper/rpc/RPC.java]: Very nit: `ArrayList<I> inputs = new ArrayList<>();` suffices

File: containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala]: Clean up imports.

File: containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-rpc/src/test/scala/ai/clipper/rpc/RPCEventHistoryTest.scala]: Remove commented, obsolete JUnit tests

File: containers/jvm/clipper-spark/pom.xml
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/pom.xml]: Not sure if commented pom was auto-generated and or will be necessary later. If not, let's remove it.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Uncomment this before committing! i.e.

`dcrankshaw/spark-scala-container` --> `clipper/spark-scala-container`

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Doc this! What is it for?

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: When documenting, be sure to include the requirement that model inherit from `MLlibModel` or `PipelineModel`

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Remove

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: We should call `publishModelToClipper` after determining whether or not `clipperHost` is local. If it is not local, we should check for the presence of ssh information before attempting to publish. We shouldn't publish a model that can't be deployed due to an absence of credentials.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:277]: The first part of the comment indicating uncertainty seems unnecessary

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:317]: We will never hit this case due to the check against `getReplOutputDir(sc).isDefined` above

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Consider defining a `createSparkContainer` method to avoid repeating code for each case

File: containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/ClipperTest.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/ClipperTest.scala]: This file is a duplicate of `containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/Clipper.scala`

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Most comments pertain to code cleanup (removing extraneous commented code, etc). Several small changes requested.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Added method documentation to this method that specifies this.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: Done. Moved it below the check for sudo credentials and below copying the model data to Clipper. We will only publish the model to Clipper if the ssh credentials were both supplied and have access to the machine.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:277]: I actually want to leave it in there, because I'm pretty sure there is a way to do it with the type system, which would be preferable. If someone (including future me) comes along and has a bright idea about how to do that, that comment may prompt them to do something.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala:317]: That's true, but Scala wants an exhaustive pattern match. And I'm leaving some commented out code in there intentionally because I hope to support deploying from the REPL in the future.

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: I think that the additional type parameter wrangling will actually make that less readable

File: containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/ClipperTest.scala
Comment (dcrankshaw) [reply containers/jvm/clipper-spark/src/test/scala/ai/clipper/spark/ClipperTest.scala]: Ahh good catch

**Comment by dcrankshaw** [comment]
@Corey-Zumar I addressed the review comments. Take another look if you feel like it but hold off on merging for now. I'd like to do a bit more testing and add a little more documentation tomorrow before we merge.

**Comment by dcrankshaw** [comment]
@Corey-Zumar Okay this is ready to go I think

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by dcrankshaw** [comment]
jenkins test this please

File: containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala
Comment (Corey-Zumar) [inline containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala]: "The model container" instead of "This model container"

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
Looks good! Just one small documentation change to avoid potential confusion

**Comment by dcrankshaw** [comment]
Addressed

**Review by Corey-Zumar — APPROVED** [review]
LGTM

**Pull request merged**
